frontend/lib/js/category-trees/reducer.js

```diff
@@ -10,24 +10,113 @@ import {
   LOAD_TREE_SUCCESS,
   LOAD_TREE_ERROR,
   CLEAR_TREES,
-}                       from './actionTypes';
+  SEARCH_TREES_START,
+  SEARCH_TREES_END,
+}                                         from './actionTypes';
 
-import { setTree }      from './globalTreeStoreHelper';
+import { setTree, getConceptById }        from './globalTreeStoreHelper';
 
 export type TreesType = { [treeId: string]: NodeType }
 
+export type SearchType = {
+  searching: boolean,
+  searchStr: string,
+  words: Array<string>,
+  result: TreesType
+}
+
 export type StateType = {
   loading: boolean,
   version: any,
-  trees: TreesType
+  trees: TreesType,
+  search?: SearchType,
 };
 
 const initialState: StateType = {
   loading: false,
   version: null,
-  trees: {}
+  trees: {},
+  search: { searching: false, words: [], result: {} }
 };
 
+const searchTreesEnd = (state: StateType, action: Object): StateType => {
+  return {
+    ...state
+  }
+}
+
+const searchTreesStart = (state: StateType, action: Object): StateType => {
+  const { searchStr } = action.payload;
+  const searching = searchStr.length > 2;
+
+return {
+    ...state,
+    search: {
+      searching: searching,
+      searchStr: searchStr,
+      words: searchStr ? searchStr.split(' ') : [],
+      result: searching ? searchingTrees(state.trees, searchStr) : {}
+    }
+  }
+}
+
+const searchingTrees = (trees: TreesType, searchStr) => {
+  const regex = new RegExp(searchStr
+    .replace(/[\\[\]\\{}()+*?.$^|]/g,
+    function (match) { return '\\' + match }), "i");
+
+  return Object.assign({}, ...Object.entries(trees).map(([treeId, treeNode]) => ({
+    [treeId]: findTreeNodes(treeId, treeNode, regex)
+  })).filter(r => { return Object.keys(r).some(k => r[k].length > 0) }));
+}
+
+const findTreeNodes = (treeId: string, treeNode: NodeType, searchStr: string) => {
+  if (treeNode === null)
+    treeNode = getConceptById(treeId);
+
+  const children = treeNode.children || [];
+  const result = children.map(child => findTreeNodes(child, null, searchStr))
+    .reduce((agg, cur) => [...agg, ...cur], []);
+
+  const label = treeNode.label || '';
+  const description = treeNode.description || '';
+  const additionalInfos = treeNode.additionalInfos
+    ? treeNode.additionalInfos.map(t => { return t.key + " " + t.value   }).join('')
```

Comment:
This can be `treeNode.additionalInfos.map(t => t.key + " " + t.value)`, or even ``treeNode.additionalInfos.map(({key, value}) => `${key} ${value}`)``.

Shouldn't we be doing `.join(' ')`?